### PR TITLE
fix: Fix linearity checking bug

### LINKED ISSUE #438: Linearity checking bug
The following program incorrectly fails linearity checking:

```python
@guppy.struct(module)
class MyStruct:
    q: qubit


@guppy.declare(module)
def use(q: qubit) -> None: ...


@guppy(module)
def test(s: MyStruct) -> MyStruct:
    use(s.q)
    while True:
        s.q = qubit()
        return s
    # Guppy is not yet smart enough to detect that this code is unreachable
    s.q = qubit()
    return s
```

### PR DESCRIPTION
**Pull request by mark-koch**
Fixes #438.

When checking that all live linear places in scope are definitely used, we have to be more careful about determining whether variables are *actually* live rather than only being included in scope because of the less precise first pass of liveness analysis.

### CODE REVIEW

File: tests/integration/test_inout.py [hunk 1]
```diff
@@ -266,7 +266,6 @@ def main(s: MyStruct, t: MyStruct) -> MyStruct:
     validate(module.compile())
 
 
-@pytest.mark.skip("Fails due to https://github.com/CQCL/guppylang/issues/337")
```
Comment (mark-koch) [inline tests/integration/test_inout.py:269]: Drive-by: This test works now

**Review by ss2165 — APPROVED** [review]
reassuring fix!
